arroyo/processing/strategies/decoder.py

```
@@ -0,0 +1,130 @@
+            initializer,
+        )
+
+    def poll(self) -> None:
+        self.__task_runner.poll()
+
+    def submit(self, message: Message[KafkaPayload]) -> None:
+        self.__task_runner.submit(message)
+
+    def close(self) -> None:
+        self.__task_runner.close()
+
+    def terminate(self) -> None:
+        self.__task_runner.terminate()
+
+    def join(self, timeout: Optional[float] = None) -> None:
+        self.__task_runner.join(timeout)
+
+
+class ValidationError(Exception):
+    """
+    Placeholder. May eventually be replaced by a DLQ specific exception
+    once that feature is finalized.
+    """
+
+    pass
+
+
+class Codec(ABC, Generic[T]):
+    @abstractmethod
+    def decode(self, raw_data: bytes, validate: bool) -> T:
+        """
+        Decode bytes from Kafka message.
+        If validate is true, the validation is run.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def validate(self, data: T) -> None:
+        """
+        Runs the validation. Raises a ValidationError if the data is not valid.
+        """
+        raise NotImplementedError
+
+
+class JsonCodec(Codec[object]):
+    def __init__(self, json_schema: object) -> None:
+        self.__json_schema = json_schema
+        # Initially set to none as the validate function is not pickleable
+        self.__validate_func: Optional[Callable[[object], None]] = None
+
+    def __validate(self, data: object) -> None:
+        if self.__validate_func is None:
+            self.__validate_func = fastjsonschema.compile(self.__json_schema)
+
+        self.__validate_func(data)
+
+    def decode(self, raw_data: bytes, validate: bool) -> object:
+        decoded = rapidjson.loads(raw_data)
+        if validate:
+            self.validate(decoded)
+        return decoded
+
+    def validate(self, data: object) -> None:
+        try:
+            self.__validate(data)
+        except Exception as exc:
+            raise ValidationError from exc
```

Other problem with the parallel consumer. Actually we should definitively not use it here.
If a single message is bad in the parallel processing the whole batch is thrown away. I doubt this is what this is meant to do. 

Taking a step back. What do you plan to do with the ValidationError ? Should you raise an InvalidMessage instead that can be captured by the DLQ? The parallel transform in that case would let the good messages go on.